pennylane/_qubit_device.py

```diff
@@ -688,3 +697,96 @@ def sample(self, observable):
+        method, but has a lower time overhead and a similar memory overhead.
+
+        .. note::
+            The adjoint differentation method has the following restrictions:
+
+            * As it requires knowledge of the statevector, only statevector simulator devices can be
+              used.
+
+            * Only expectation values are supported as measurements.
+
+        Args:
+            tape (.QuantumTape): circuit that the function takes the gradient of
+
+        Returns:
+            np.ndarray: the derivative of the tape with respect to trainable parameters.
+            Dimensions are ``(len(observables), len(trainable_params))``.
+
+        Raises:
+            QuantumFunctionError: if the input tape has measurements that are not expectation values
+                or contains a multi-parameter operation aside from :class:`~.Rot`
+        """
+
+        for m in tape.measurements:
+            if m.return_type is not qml.operation.Expectation:
+                raise qml.QuantumFunctionError(
+                    "Adjoint differentiation method does not support"
+                    f" measurement {m.return_type.value}"
+                )
+
+            if not hasattr(m.obs, "base_name"):
+                m.obs.base_name = None  # This is needed for when the observable is a tensor product
+
+        # Perform the forward pass
+        self.reset()
+
+        # Consider using caching and calling lower-level functionality. We just need the state
+        # without postprocessing https://github.com/PennyLaneAI/pennylane/pull/1032/files#r563441040
+        self.execute(tape)
+
+        phi = self._reshape(self.state, [2] * self.num_wires)
+
+        lambdas = [self._apply_operation(phi, obs) for obs in tape.observables]
+
+        expanded_ops = []
+        for op in reversed(tape.operations):
+            if op.num_params > 1:
+                if isinstance(op, qml.Rot) and not op.inverse:
+                    ops = op.decomposition(*op.parameters, wires=op.wires)
+                    expanded_ops.extend(reversed(ops))
+                else:
+                    raise QuantumFunctionError(
+                        f"The {op.name} operation is not supported using "
+                        'the "adjoint" differentiation method'
+                    )
+            else:
+                if op.name not in ("QubitStateVector", "BasisState"):
+                    expanded_ops.append(op)
+
+        jac = np.zeros((len(tape.observables), len(tape.trainable_params)))
+        dot_product_real = lambda a, b: self._real(sum(self._conj(a) * b))
+
+        param_number = len(tape._par_info) - 1  # pylint: disable=protected-access
+        trainable_param_number = len(tape.trainable_params) - 1
+        for op in expanded_ops:
+
+            if (op.grad_method is not None) and (param_number in tape.trainable_params):
+                d_op_matrix = operation_derivative(op)
+
+            op.inv()
```

Comment:
Oh, this both returns the inverse _and_ changes `op` to its inverse. That's a bit confusing.

Reply:
Yes, I can testify that this confused me a lot! It would be nice to have a `qml.inverse` function.

Reply:
Yes, it would be nice to have a purely functional way of doing this instead.